### DISCUSSION TITLE
Modelling of the effect of a thin metal rod on electrical resisitivity measurement

### DISCUSSION THREAD

**Discussion by c-finco** [gimli-org/pyGIMLi#483] [Q&A] [answered]
Hello,

I am currently trying to use pygimli for a forward modeling issue. I have a electrode at the end of a thin metal rod (diameter of the rod is 2cm) at a depth of 1m and a second electrode on the surface. I would like to simulate the response and look at the sensitivity of a pole-pole measurement using these to electrodes. At this point I am only looking at 2-D. The pygimli version I use is 1.3.0
To this end, I am using the forward model in pygimli.physics.ERTModelling() and my code (inis heavily inspired from both the example "4-point sensitivity" and the answer you gave to issue #320.

With the rod and the pole-pole scheme, I get an apparent resistivity value below 10 ohm.m which is much lower than what I was expecting and I have problems with the mesh close to the rod :
![image](https://user-images.githubusercontent.com/72929506/214022663-f265ab82-7b27-4409-ab1c-e70ca26d3222.png)
Some intermediary tests show that everything seems to work well for an homogeneous medium :
![image](https://user-images.githubusercontent.com/72929506/214023658-f872d877-3ef8-4cd8-ade7-712d8d1fd905.png)
And when using a four electrodes scheme like dipole-dipole, the apparent resistivity value is much higher in the range I was expecting but the mesh issue is naturally still here.
![DD_withrod](https://user-images.githubusercontent.com/72929506/214023993-daf39e60-9cc3-4f8c-bc88-2cfce2a02b2c.png)

Therefore, I have 2 questions: 

1. Do you have a recommendation to modify the mesh to improve the results close the rod. Do you think refining the mesh of the homogeneous medium around the rod would help ?
2. I did not find any mention of remote electrodes in the documentation (if it exists and I missed it, I would be grateful for a link) but chose to use -1 after looking at the results of the ert.createData function with schemeName='pp'. However when looking at the position of electrode -1 I get: 
```
scheme.sensorPosition(-1)
RVector3: (0.0, 0.0, 1.5569868787319047e-231)
```
which is close to 0 and leads me to wonder if -1 is the right way to indicate a remote electrode.

Thanks a lot in advance,
Cheers,
Cécile

Here is my code:
```
#%% Modules import
import numpy as np
import matplotlib.pyplot as plt
import pygimli as pg
import pygimli.meshtools as mt
from pygimli.physics import ert

print(pg.__version__)

#%% ================= fonctions d'affichage ==================== #
def getABMN(scheme, idx):
    """Get coordinates of four-point cfg with id `idx` from DataContainerERT
    `scheme`."""
    coords = {}
    for elec in "abmn":
        elec_id = int(scheme(elec)[idx])
        elec_pos = scheme.sensorPosition(elec_id)
        coords[elec] = elec_pos.x(), elec_pos.y()
    return coords


def plotABMN(ax, scheme, idx):
    """Visualize two-point configuration on given axes.
    Modified for Pole-pole configuration"""
    coords = getABMN(scheme, idx)
    for elec in coords:
        if scheme(elec) != [-1]:
            x, y = coords[elec]
            if elec in "ab":
                color = "green"
            else:
                color = "blue"
            ax.plot(x, y, marker=".", color=color, ms=10)
            ax.annotate(
                elec.upper(),
                xy=(x, y),
                ha="center",
                fontsize=10,
                bbox=dict(boxstyle="round", fc=(0.8, 0.8, 0.8), ec=color),
                xytext=(-10, 20),
                textcoords="offset points",
                arrowprops=dict(
                    arrowstyle="wedge, tail_width=.5",
                    fc=color,
                    ec=color,
                    patchA=None,
                    alpha=0.45,
                ),
            )
            ax.plot(
                coords["a"][0],
            )


#%%  ================ Creating scheme  ===============
with_rod = True
nelecs = 4
pos = np.zeros((nelecs, 2))
pos[:, 0] = [50, 51, 52, 53]
l_rod = 1  # length of rod in the ground
pos[0, 1] = -l_rod

scheme = pg.DataContainerERT()
scheme.setSensorPositions(pos)
measurements = np.array([0, -1, 1, -1])  # Pôle-pôle avec B et N à l'infini
for i, elec in enumerate("abmn"):
    scheme[elec] = [measurements[i]]

# Geometric factor calculation
scheme["k"] = ert.createGeometricFactors(scheme)

#%% ======================= Medium creation =========================
# Modèle homogène
domain = mt.createWorld(
    start=[20, -10], end=[80, 0], marker=2, worldMarker=True, area=0.1
)
# rod in the ground
rod = mt.createRectangle(start=[49.99, -l_rod], end=[50.01, 0], marker=3)

if with_rod:
    world = domain + rod
else:
    world = domain

# Nodes at electrodes positions
for pos in scheme.sensorPositions():
    world.createNode(pos)
    world.createNode(pos - pg.RVector3(0, 0.05))

# Mesh creation
mesh = mt.createMesh(world, quality=33)
pg.show(mesh, showMesh=True, markers=True)
mesh.exportVTK("mesh_2D.vtk")

# Resistivity model
rhomap = [[1, 100.0], [2, 100.0], [3, 0.001]]
rhomodel = pg.solver.parseArgToArray(rhomap, mesh.cellCount(), mesh)

# Append triangulated boundary to the mesh
bigMesh = mt.appendTriangleBoundary(mesh, xbound=250, ybound=250, marker=1)
rhomodel2 = pg.solver.parseArgToArray(rhomap, bigMesh.cellCount(), bigMesh)
pg.show(bigMesh, showMesh=True, markers=True)
bigMesh.exportBoundaryVTU("boundary.vtu")
print("Mesh created")

#%% ===================== Simulation and sensitivity estimation ===================
fop = ert.ERTModelling()
fop.setData(scheme)
fop.setMesh(bigMesh)
fop.setRegionProperties(1, background=False)

resp = fop.response(rhomodel2)
print(f"Modeled electrical resistivity : {resp[0]} ohm.m")

fop.createJacobian(rhomodel2)
J = fop.jacobian()

#%% ================= Affichage ===============
# ax, cb = pg.show(fop.paraDomain, J[0, :], cMin=-0.05, cMax=0.05, cMap="bwr")
# ax.set_xlim([45, 60])
# ax.set_ylim([-5, 0])
#

labels = [f"$\\rho_a$ = {np.round(resp[0])} $\\Omega$.m"]
fig, ax = plt.subplots(1, 1, sharex=True, figsize=(8, 4))

for i, sens in enumerate(fop.jacobian()):
    # Label in lower-left corner
    ax.text(
        0.01,
        0.15,
        labels[i],
        horizontalalignment="left",
        verticalalignment="top",
        transform=ax.transAxes,
        fontsize=12,
        fontweight="bold",
    )

    # Electrode annotations
    plotABMN(ax, scheme, i)

    # Log-scaled and normalized sensitivity
    normsens = pg.utils.logDropTol(sens / fop.paraDomain.cellSizes())
    # normsens = sens / fop.paraDomain.cellSizes()
    normsens /= np.max(normsens)
    pg.show(
        fop.paraDomain,
        normsens[fop.paraDomain.cellMarkers()],
        cMap="RdGy_r",
        ax=ax,
        orientation="vertical",
        label="Normalized\nsensitivity",
        nLevs=3,
        cMin=-1,
        cMax=1,
    )
ax.set_ylim([-3, 0])
ax.set_xlim([45, 55])
plt.tight_layout()
plt.show()

# %%
```

**Comment by c-finco** [comment]
Hi,

An update on my issue.
My main issue was with the geometric factor. Because one electrode is not on the surface I needed to use the numerical=True option with the CreateGeometricFactors function.
Along with refining the mesh, this solved my issue.

Cheers,
Cécile

**Comment by halbmy** [comment, marked as answer]
Sorry for not responding earlier. A few comments:
1. Yes, remote electrodes are defined by -1 in the ABMN tokens (0 in data files). But they are infinite by definition, so they don't appear in the electrode list. If they had positions, the would be normal electrodes, just with a remote position. `scheme.sensorPosition(-1)` should provide the last one, and `scheme.sensorPosition(-2)` the last-but-one etc.
2. For modelling extended line electrodes you could better use the shunt electrode model (SEM) as defined by Ronczka et al. (2015) based on Rücker & Günther (2011). See Figure 6 for sensitivity with long electrodes. It can be easily achieved by creating nodes for the rod with marker -10000 (for the first SEM, -10001 for the second etc.). However, i never checked that in 2D, would be interesting to see. 

Ronczka, M., Rücker, C. & Günther, T. (2015): Numerical study of long electrode electric resistivity tomography – Accuracy, sensitivity and resolution. Geophysics 80(6), E317-328, doi:10.1190/geo2014-0551.1.
Rücker, C. & Günther, T. (2011): The simulation of Finite ERT electrodes using the complete electrode model. Geophysics 76(4), F227-238, doi:10.1190/1.3581356.
